### ISSUE TITLE
Android Example App for real-time speech-to-text

### ISSUE THREAD

**Issue by OscarVanL** [TensorSpeech/TensorFlowASR#35]
Hi,

I'm evaluating different techniques for creating speech recognition models that are suitable for use on an Android app.

I'm looking for a way to perform speech-to-text in a real-time, streaming fashion, which is performant even on an Android phone. The full plan is for the output of the speech-to-text to enter a text-to-speech (FastSpeech2 + MB-MelGAN) model for speech synthesis with as little latency as possible.

I see in this library you have implemented a few models, and support exporting models to `tflite`. Which would you recommend for use on Android?

Finally, do you have any examples using this project in an Android app? In the other TensorSpeech project (TensorFlowTTS), there is an example, and it does not look trivial to bring such a complex project to Android.

Thanks!

**Comment by usimarit** [comment]
I don't have any examples for Android because I haven't had time for that.

In my view, the best choice for **streaming** speech recognition is the paper: [https://arxiv.org/abs/1811.06621](https://arxiv.org/abs/1811.06621). I implemented ONLY the model, other techniques in the paper are not yet supported, and I haven't trained it yet.

The reason why I choose that model over other SOTA models is because there're states in the encoder passed from previous chunk to next chunk, which avoids losing previous audio features information.

**Comment by mapledxf** [comment]
@usimarit  do u have the pretrained model? maybe i can do the android demo based on the pretrain model.

**Comment by OscarVanL** [comment]
@mapledxf That would be awesome, your TensorFlowTTS Android example is excellent :)

**Comment by usimarit** [comment]
@mapledxf I ONLY have conformer pretrained model, which is unusable when applying for **streaming**, but it works well for **non-streaming**. It would be great if you can add Android demo for Conformer :smile:

**Comment by alsm168** [comment]
the conformer can be transformed to be  **streaming** if add mask to mask the future information in transducer encoder?

**Comment by usimarit** [comment]
> the conformer can be transformed to be  **streaming** if add mask to mask the future information in transducer encoder?

Maybe, but it still loses audio features information of the previous chunks because there're no states in the encoder.

**Comment by mapledxf** [comment]
@usimarit  I have already tried to deploy the conformer model to the android device. The problem is it is slow and it doesnt support streaming. :(

**Comment by usimarit** [comment]
> @usimarit  I have already tried to deploy the conformer model to the android device. The problem is it is slow and it doesnt support streaming. :(

I deployed the conformer on my website, and it's kinda fast. How is it slow? Like 3-4s latency?

**Comment by mapledxf** [comment]
> > @usimarit  I have already tried to deploy the conformer model to the android device. The problem is it is slow and it doesnt support streaming. :(
> 
> I deployed the conformer on my website, and it's kinda fast. How is it slow? Like 3-4s latency?

yeah, it's like 3 sec delay on my android pad.

**Comment by usimarit** [comment]
@mapledxf on my website, excluding the delay of the network, it's like 1s latency.

**Comment by mapledxf** [comment]
> @mapledxf on my website, excluding the delay of the network, it's like 1s latency.

The mobile's performance is much lower than the pc or server because of the cpu architecture's differences.

**Comment by OscarVanL** [comment]
Thank you for checking this for me mapledxf. Is it possible that other model types would be faster on Android?

**Comment by usimarit** [comment]
> Thank you for checking this for me mapledxf. Is it possible that other model types would be faster on Android?

I believe **streaming** models can be faster because they predict on audio chunk, which is 4-5 times smaller than the whole audio, and they predict while we're speaking, whereas **non-streaming** only predict after we speak.

**Comment by mapledxf** [comment]
need a pretrained streaming model :(

**Comment by mapledxf** [comment]
@usimarit I'm trying to train a streaming transducer model with LibriSpeeach dataset.
I just changed the dataset path in the config file.

bug the loss looks very odd:

<img width="336" alt="image" src="https://user-images.githubusercontent.com/3639832/97267196-b9ec4f00-1864-11eb-83e1-a6a90abb58ee.png">

Any idea why?

**Comment by usimarit** [comment]
> @usimarit I'm trying to train a streaming transducer model with LibriSpeeach dataset.
> I just changed the dataset path in the config file.
> 
> bug the loss looks very odd:
> 
> <img alt="image" width="379" src="https://user-images.githubusercontent.com/3639832/97267347-fb7cfa00-1864-11eb-823e-d8152daffe7a.png">
> 
> Any idea why?

I think it's stuck at local minima, try removing augmentations.

**Comment by mapledxf** [comment]
I removed the augmentation in the config.yaml, and still got the same result:
This is the config.yml:
[config.txt](https://github.com/TensorSpeech/TensorFlowASR/files/5449176/config.txt)

This is the training log:
<img width="814" alt="image" src="https://user-images.githubusercontent.com/3639832/97383525-cc1fc900-1908-11eb-8e89-284e49cb63c5.png">

**Comment by usimarit** [comment]
> I removed the augmentation in the config.yaml, and still got the same result:
> This is the config.yml:
> [config.txt](https://github.com/TensorSpeech/TensorFlowASR/files/5449176/config.txt)
> 
> This is the training log:
> <img alt="image" width="814" src="https://user-images.githubusercontent.com/3639832/97383525-cc1fc900-1908-11eb-8e89-284e49cb63c5.png">

I haven't added the learning rate to the optimizer, try use learning rate 0.0001 or change the optimizer to SGD

**Comment by mapledxf** [comment]
@usimarit 
I've changed the optimizer_config in config.yaml to the following, but still no luck. :(
```
  optimizer_config:
    class_name: adam
    config:
      learning_rate: 0.0001
```

**Comment by mapledxf** [comment]
I changed 
```
    optimizer = tf.keras.optimizers.Adam(learning_rate=0.0001)
```
in the train_streaming_transducer.py seems to work now. :)

**Comment by mapledxf** [comment]
@OscarVanL you can close this issue for now.

I will create a PR once I finished training and the Android Demo if everything works fine.

**Comment by usimarit** [comment]
> I changed
> 
> ```
>     optimizer = tf.keras.optimizers.Adam(learning_rate=0.0001)
> ```
> 
> in the train_streaming_transducer.py seems to work now. :)

Yeah, the Adam Optimizer needs small learning rate or a scheduler to work well :smile:

**Comment by OscarVanL** [comment]
> @OscarVanL you can close this issue for now.
> 
> I will create a PR once I finished training and the Android Demo if everything works fine.

Awesome. Thank you for all the effort you're putting into this. I look forward to it, if you get it working :)
